Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift

```diff
@@ -73,43 +87,57 @@ extension BTPaymentFlowClient: BTPaymentFlowClientDelegate {
 
     public func onPayment(with url: URL?, error: Error?) {
         if let error {
-            _apiClient.sendAnalyticsEvent("ios.\(paymentFlowName).start-payment.failed")
+            _apiClient.sendAnalyticsEvent(BTPaymentFlowAnalytics.paymentFailed)
             onPaymentComplete(nil, error: error)
             return
         }
         
         guard let url else {
+            _apiClient.sendAnalyticsEvent(BTPaymentFlowAnalytics.paymentFailed)
             onPaymentComplete(nil, error: BTPaymentFlowError.missingRedirectURL)
             return
         }
         
-        _apiClient.sendAnalyticsEvent("ios.\(paymentFlowName).webswitch.initiate.succeeded")
-        
-        authenticationSession = ASWebAuthenticationSession(url: url, callbackURLScheme: BTCoreConstants.callbackURLScheme, completionHandler: { callbackURL, error in
-            // Required to avoid memory leak for BTPaymentFlowClient
-            self.authenticationSession = nil
-            
-            // TODO: - Refactor similar to BTPayPalClient to handle distinct cancellations b/w system alert or system browser
-            if let error = error as? NSError {
-                if error.domain == ASWebAuthenticationSessionError.errorDomain,
-                   error.code == ASWebAuthenticationSessionError.canceledLogin.rawValue {
-                    self._apiClient.sendAnalyticsEvent("ios.\(self.paymentFlowName)).authsession.browser.cancel")
+        returnedToAppAfterPermissionAlert = false
+        webAuthenticationSession.start(
+            url: url,
+            context: self,
+            sessionDidDisplay: { [weak self] didDisplay in
+                if didDisplay {
+                    self?._apiClient.sendAnalyticsEvent(BTPaymentFlowAnalytics.browserPresentationSucceeded)
+                } else {
+                    self?._apiClient.sendAnalyticsEvent(BTPaymentFlowAnalytics.browserPresentationFailed)
```

Comment:
If it's really important to be able to send `webSessionFailedToLaunch` error back to merchant at end of the error block, we can set boolean flag here and reset before calling the webSession.

Reply:
I don't see the error `webSessionFailedToLaunch` anywhere in this codebase currently aside from in the CHANGELOG and error enum. Would you mind clarifying where this error is being sent?

Reply:
I see it  in our feature branch I think, on BTPayPalError.

Reply:
I did have it in previous commit but @scannillo thought that we shouldn't return to merchant app at this point with an error but it should be handled in completion handler. In PayPal module, we did return with completion with error on failure to launch and sent that webSessionFailedToLaunch. So I was wondering what to do with that error. How we make it consistent. Sorry, it took some digging to last week me.

Reply:
In PayPal it looks like we send `webSessionFailedToLaunch` if `isAuthenticationSessionStarted = false`. It looks like we don't use `isAuthenticationSessionStarted` in `PaymentFlow` at all so I don't know that we need to add it.

I think for now we can go with what we have but should have a follow up PR to align how we are using `WebAuthenticationSession` as a whole. We shouldn't need 3 different implementations of `WebAuthenticationSession`. I think this brings us back to an earlier discussion we had around having a shared `WebAuthenticationSession` in Core that we use across all modules since we have deviated and have 3 different implementations in SEPA vs PayPal vs PaymentFlow. Even if we don't move it into Core we likely can/should use the same implementation across the modules unless there is a compelling reason not to.

Reply:
➕ 1

Reply:
Victoria - we don't want to send a tokenization fail event for the bool status of the ASWeb. That is because we are not calling the merchant's completion in that case. I don't think that bool value is reliable enough.

There is a discrepancy with the PayPal implementation [here](https://github.com/braintree/braintree_ios/blob/9cb3aa0b50b027e63f93a8abc4cda6492d897b1a/Sources/BraintreePayPal/BTPayPalClient.swift#L357). This PayPal code linked is incorrect and needs to be updated.

Not calling a merchant completion nor sending a tokenize failed event for the `asWeb.start()` bool result matches our [v5 implementation](https://github.com/braintree/braintree_ios/blob/8f34d3f1deada3559ad287d41d0fb7c65de2c9b3/Sources/BraintreePayPal/BTPayPalDriver.m#L299-L305).

Reply:
The reason I made separate implementation here was that the PayPal implementation did not have separate completion for` sessionDidDisplay` and `sessionDidComplete`. At the time I did implementation for PayPal, I was not aware that we were conforming to PPCP way of doing this.

 I didn't think it would matter since if not failed here with merchant completion here, it will be caught in the completion as it is with the implementation without sending merchant completion fail here. I do agree that we want consistency but I think either way, failure to display returns to merchant app with fail eventually.

Reply:
Here are questions I have remaining: 
1> Do we want to send networkConnectionLost analytics for all the modules? Sammy, I think for one of the modules, we decide not to and I had a lapse for LocalPayment module and included it.
2> Should I make a separate ticket for not returning to merchant completion right away with sessionDidDisplay false in PayPal module.?
3> Do we make a separate ticket for unified websession module?
Do we want to include refactoring PayPal module to use this wrapper class for ASWebSession instead of using ASWebSession directly?

Reply:
> 1> Do we want to send networkConnectionLost analytics for all the modules? Sammy, I think for one of the modules, we decide not to and I had a lapse for LocalPayment module and included it.

I think this is up to us to decide. The merchant that requested this to be added only uses PayPal and Venmo through us. It may make sense to add it to the missing modules in a follow up PR as it looks like we have it in about half the modules now.

> 2> Should I make a separate ticket for not returning to merchant completion right away with sessionDidDisplay false in PayPal module?

We should make a follow up PR as a whole to align our usage of ASWebAuthenticationSession as a whole. All of our modules depend on Core so I think it makes sense to move the implementation of our WebAuthenticationSession into there in a follow up PR.

> 3> Do we make a separate ticket for unified websession module?
Do we want to include refactoring PayPal module to use this wrapper class for ASWebSession instead of using ASWebSession directly?

Yes, this should be in a follow up PR and not in this PR.

Reply:
Thank you Jax. So should those be two separate PR's - one for not returning to merchant fail completion after failure to display in PayPal module, and one for refactoring PayPal modulate make unified WebSession class?

Reply:
And networkConnectionLost, if we decide to include, we should have another PR to add them all in.

Reply:
> Thank you Jax. So should those be two separate PR's - one for not returning to merchant fail completion after failure to display in PayPal module, and one for refactoring PayPal module to use wrapper class for web session and make unified WebSession class for all three modules?

Yep, we should have a follow up PR to align our usage and implementation of ASWebAuthenticationSession.

Reply:
Resolving: ticket for unified ASWeb logic is DTBTSDK-2712